kong/observability/logs.lua

```diff
@@ -0,0 +1,160 @@
+local _M = {
+  maybe_push = function() end,
+  get_request_logs = function() return {} end,
+  get_worker_logs = function() return {} end,
+}
+
+if ngx.config.subsystem ~= "http" then
+  return _M
+end
+
+
+local request_id_get = require "kong.observability.tracing.request_id".get
+local time_ns = require "kong.tools.time".time_ns
+local cycle_aware_deep_copy = require "kong.tools.utils".cycle_aware_deep_copy
+
+local get_log_level = require "resty.kong.log".get_log_level
+local constants_log_levels = require "kong.constants".LOG_LEVELS
+
+local table_new = require "table.new"
+local table_clear = require "table.clear"
+local string_buffer = require "string.buffer"
+
+local ngx = ngx
+local kong = kong
+local table = table
+local tostring = tostring
+
+local MAX_WORKER_LOGS = 1000
+local MAX_REQUEST_LOGS = 1000
+local NGX_CTX_REQUEST_LOGS_KEY = "o11y_logs_request_scoped"
+
+local worker_logs = table_new(100, 0)
+local logline_buf = string_buffer.new()
+
+
+local function configured_log_level()
+  local ok, level = pcall(get_log_level)
+  if not ok then
+    level = constants_log_levels[kong.configuration.log_level]
+  end
+
+  return level
+end
+
+
+-- needed because table.concat doesn't like booleans
+local function concat_tostring(tab)
+  if #tab == 0 then
+    return ""
+  end
+
+  for i = 1, #tab do
+    logline_buf:put(tostring(tab[i]))
```

Comment:
> Lua `nil` arguments are accepted and result in literal `"nil"` string while Lua booleans result in literal `"true"` or `"false"` string outputs. And the `ngx.null` constant will yield the `"null"` string output.
> https://github.com/openresty/lua-nginx-module?tab=readme-ov-file#ngxlog

We should keep the consistency of the log lines produced by `ngx.log`, which only accepts:

* Lua number
* Lua string
* Lua boolean
* Lua table with the metamethod `__tostring` ([Ref](https://github.com/openresty/lua-nginx-module/blob/6477a7b46d78e9ad841f2f3b2e1d37161b748669/src/ngx_http_lua_log.c#L169-L174))
* `ngx.null`

Other types will raise a Lua error except for non-null `userdata`.